app/bundles/LeadBundle/EventListener/ReportDNCSubscriber.php

```diff
@@ -0,0 +1,212 @@
+
+        if ($event->hasColumn(['u.first_name', 'u.last_name']) || $event->hasFilter(['u.first_name', 'u.last_name'])) {
+            $qb->leftJoin('l', MAUTIC_TABLE_PREFIX.'users', 'u', 'u.id = l.owner_id');
+        }
+
+        if ($event->hasColumn('i.ip_address') || $event->hasFilter('i.ip_address')) {
+            $event->addLeadIpAddressLeftJoin($qb);
+        }
+
+        if ($this->companyReportData->eventHasCompanyColumns($event)) {
+            $qb->leftJoin('l', MAUTIC_TABLE_PREFIX.'companies_leads', 'companies_lead', 'l.id = companies_lead.lead_id');
+            $qb->leftJoin('companies_lead', MAUTIC_TABLE_PREFIX.'companies', 'comp', 'companies_lead.company_id = comp.id');
+        }
+
+        $event->setQueryBuilder($qb);
+    }
+
+    /**
+     * @param ReportDataEvent $event
+     */
+    public function onReportDisplay(ReportDataEvent $event)
+    {
+        if (!$event->checkContext([self::DNC])) {
+            return;
+        }
+
+        $data = $event->getData();
+
+        if (isset($data[0]['reason']) || isset($data[0]['channel']) || isset($data[0]['channel_id'])) {
+            foreach ($data as $key => &$row) {
+                if (isset($row['reason'])) {
+                    $row['reason'] = $this->getDncReasonLabel($row['reason']);
+                }
+
+                if (isset($row['channel']) && $row['channel_id']) {
+                    $href              = $this->router->generate('mautic_'.$row['channel'].'_action', ['objectAction' => 'view', 'objectId' => $row['channel_id']]);
+                    $row['channel']    = '<a href="'.$href.'" data-toggle="ajax">'.$this->channelListHelper->getChannelLabel($row['channel']).'</a>';
+                    $row['channel_id'] = '<a href="'.$href.'" data-toggle="ajax">'.$row['channel_id'].'</a>';
+                }
+                if (isset($row['channel'])) {
+                    $row['channel'] = $this->channelListHelper->getChannelLabel($row['channel']);
+                }
+            }
+
+            $event->setData($data);
+        }
+    }
+
+    /**
+     * @return array
+     */
+    private function getDncReasons()
+    {
+        return [
+            0 => $this->translator->trans('mautic.lead.report.dnc_contactable'),
+            1 => $this->translator->trans('mautic.lead.report.dnc_unsubscribed'),
+            2 => $this->translator->trans('mautic.lead.report.dnc_bounced'),
+            3 => $this->translator->trans('mautic.lead.report.dnc_manual'),
+        ];
+    }
+
+    /**
+     * @param int $reasonId
+     *
+     * @return string
+     *
+     * @throws \UnexpectedValueException
+     */
+    private function getDncReasonLabel($reasonId)
+    {
+        if (isset($this->getDncReasons()[$reasonId])) {
+            return $this->getDncReasons()[$reasonId];
+        }
+
+        throw new \UnexpectedValueException("There is no DNC reason with ID '{$reasonId}'");
+    }
```

Comment:
If we need to add DNC reasons beyond what was initially scoped (for instance new GB rules, National registries, etc) this behavior means having to rewrite core in several places. Should we have a more dynamic way of managing reason consts => strings than throwing exceptions when the original scope isnt found? It kinda paints the whole DNC functionality into a corner and prevents extension without massive core updates.

Reply:
One solution would be to create a repo method on DoNotContact entity that returns the originally scoped list plus an event dispatch that allows plugins to add DNC reasons and translation strings. Then we only ever need one global method to get a list of acceptable reasons and toText strings.

Reply:
I don't see anything wrong in this method. The problem you have is in the method `getDncReasons` above where the DNC reasons are hard-coded.

I suggest to create new service, say `DncReasonProvider` which would set the 4 core reasons and dispatched an event to add more. Then inject the provider into this subscriber and get the reasons from it.

@scottshipman could you send a PR against this PR/branch with this new service?